### PR TITLE
[feature-mesh-scaling] [collection] Add 'instance' module

### PR DESCRIPTION
**Pull request by relrod**
##### SUMMARY

First go at awx_collection module for `instance`

##### ISSUE TYPE
 - New or Enhanced Feature

##### COMPONENT NAME

 - Collection

### CODE REVIEW

File: awx_collection/plugins/modules/instance.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        - Role that this node plays in the mesh.
+      choices:
+        - control
+        - execution
+        - hybrid
+        - hop
+      required: False
+      type: str
+      default: execution
+    node_state:
```
Comment (john-westcott-iv) [inline awx_collection/plugins/modules/instance.py:55]: Couple questions here...
1) Normally ansible just uses state ([absent, present]) do we want to call this just state? (we could alter test_completness to accept this)
2) Should the module be allowed to set all of these states? i.e. should only the controller be allowed to put something into the -fail state/ready state/etc?
3) Should there be an additional "state" parameter to allow (someday) for deletion of the endpoint (or is that never planned).

**Review by john-westcott-iv — COMMENTED** [review]
Do we need anything for downloading the zip file?

File: awx_collection/plugins/modules/instance.py (hunk 1 above)
Comment (relrod) [reply awx_collection/plugins/modules/instance.py:55]: > Normally ansible just uses state ([absent, present]) do we want to call this just state? (we could alter test_completness to accept this)

I did consider adding a `state` field, I actually had one at first and was considering doing something like making it set `node_state=deprovisioning` when `state=absent`, but I think that gets confusing pretty quickly, and as Hao told me last week, there are some things that happen during the deprovisioning process other than just deleting the node. So I'm not sure it makes sense to combine them here or to rename `node_state` as `state`. @TheRealHaoLiu Do you have thoughts on this?

> Should the module be allowed to set all of these states? i.e. should only the controller be allowed to put something into the -fail state/ready state/etc?

Probably another question for @TheRealHaoLiu - is it safe to just limit `node_state` to `installing` and `deprovisioning`?

Comment (jbradberry) [reply awx_collection/plugins/modules/instance.py:55]: > Normally ansible just uses state ([absent, present]) do we want to call this just state?

@john-westcott-iv Definitely not.  This is not the equivalent of Ansible's assertions about a package being installed on the system, but rather a value that affects how a node behaves and can be in any of a half-dozen distinct states.

Comment (jbradberry) [reply awx_collection/plugins/modules/instance.py:55]: (that is to say, this value is a model field.)

Comment (jbradberry) [reply awx_collection/plugins/modules/instance.py:55]: > Should the module be allowed to set all of these states?

No, it shouldn't.  I'd limit it, at the current time, to `installed` and `deprovisioning`.

Comment (jbradberry) [reply awx_collection/plugins/modules/instance.py:55]: `provisioning` will be brought in later.

File: awx_collection/plugins/modules/instance.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      required: False
+      type: bool
+      default: True
+    managed_by_policy:
+      description:
+        - Managed by policy
+      required: False
+      default: True
+      type: bool
+    node_type:
```
Comment (jbradberry) [inline awx_collection/plugins/modules/instance.py:47]: At the moment, we are only allowing the creation of nodes of type `execution`.

**Review by jbradberry — APPROVED** [review]
These two fields need to have their choices restricted, but this otherwise seems fine.

**Review by jbradberry — APPROVED** [review]

**Review by TheRealHaoLiu — APPROVED** [review]

**Pull request merged**